diff --git a/src/arch/x86/asm.h b/src/arch/x86/asm.h
index 81b3268..312c8aa 100644
--- a/src/arch/x86/asm.h
+++ b/src/arch/x86/asm.h
@@ -15,6 +15,14 @@ _##symbol:
 
 # define END(symbol)
 
+#elif defined(__MSYS__)
+
+# define NESTED(symbol) \
+    .globl symbol; \
+symbol:
+
+# define END(symbol)
+
 #else
 
 # define NESTED(symbol) \
diff --git a/src/kern/task/hev-task-stack-mmap.c b/src/kern/task/hev-task-stack-mmap.c
index 6008780..2b3ed04 100644
--- a/src/kern/task/hev-task-stack-mmap.c
+++ b/src/kern/task/hev-task-stack-mmap.c
@@ -46,7 +46,12 @@ hev_task_stack_new (int size)
         return NULL;
 
     self->stack = mmap (NULL, size, PROT_READ | PROT_WRITE,
-                        MAP_PRIVATE | MAP_ANON | MAP_STACK, -1, 0);
+                        MAP_PRIVATE | MAP_ANON
+#ifndef __MSYS__
+// TODO: MSYS2 does not have MAP_STACK
+                        | MAP_STACK
+#endif
+                        , -1, 0);
     if (self->stack == MAP_FAILED) {
         hev_free (self);
         return NULL;
